tests/test_syntax/blocks/test_html_blocks.py

```diff
@@ -0,0 +1,783 @@
+            '<p></p>',
+            '<p></p>'
+        )
+
+    def test_raw_empty_space(self):
+        self.assertMarkdownRenders(
+            '<p> </p>',
+            '<p> </p>'
+        )
+
+    def test_raw_empty_newline(self):
+        self.assertMarkdownRenders(
+            '<p>\n</p>',
+            '<p>\n</p>'
+        )
+
+    def test_raw_empty_blank_line(self):
+        self.assertMarkdownRenders(
+            '<p>\n\n</p>',
+            '<p>\n\n</p>'
+        )
+
+    def test_multiline_raw(self):
+        self.assertMarkdownRenders(
+            self.dedent(
+                """
+                <p>
+                    A raw paragraph
+                    with multiple lines.
+                </p>
+                """
+            ),
+            self.dedent(
+                """
+                <p>
+                    A raw paragraph
+                    with multiple lines.
+                </p>
+                """
+            )
+        )
+
+    def test_blank_lines_in_raw(self):
+        self.assertMarkdownRenders(
+            self.dedent(
+                """
+                <p>
+
+                    A raw paragraph...
+
+                    with many blank lines.
+
+                </p>
+                """
+            ),
+            self.dedent(
+                """
+                <p>
+
+                    A raw paragraph...
+
+                    with many blank lines.
+
+                </p>
+                """
+            )
+        )
+
+    def test_raw_surrounded_by_Markdown(self):
+        self.assertMarkdownRenders(
+            self.dedent(
+                """
+                Some *Markdown* text.
+
+                <p>*Raw* HTML.</p>
+
+                More *Markdown* text.
+                """
+            ),
+            self.dedent(
+                """
+                <p>Some <em>Markdown</em> text.</p>
+                <p>*Raw* HTML.</p>
+
+                <p>More <em>Markdown</em> text.</p>
+                """
+            )
+        )
+
+    def test_raw_without_blank_lines(self):
+        self.assertMarkdownRenders(
+            self.dedent(
+                """
+                Some *Markdown* text.
+                <p>*Raw* HTML.</p>
+                More *Markdown* text.
+                """
+            ),
+            # The raw gets treated as inline HTML. This follows the rules and this lib's
+            # previous behavior, but not the reference implementation. TODO: Reevaluate.
+            self.dedent(
+                """
+                <p>Some <em>Markdown</em> text.
+                <p><em>Raw</em> HTML.</p>
+                More <em>Markdown</em> text.</p>
+                """
+            )
+            # The reference implementation does this instead:
+            # self.dedent(
+            #     """
+            #     <p>Some <em>Markdown</em> text.</p>
+            #     <p>*Raw* HTML.</p>
+            #     <p>More <em>Markdown</em> text.</p>
+            #     """
```

Comment:
I think our current behavior is fine.

Reply:
Note that with the current behavior, that last line is not wrapped in a `p` tag. There is a closing tag, bu no opening tag. Conversely, the first lien has an opening tag, but no closing tag (although it gets auto-closed by the opening tag of the second line).

Reply:
Ah, I did not notice that. Then maybe we should do the same thing as reference implementation, as valid output is better than invalid.